File: CICStatus.py
#!/usr/bin/env python

import urllib2, urllib, re, smtplib, datetime, time, random, cookielib
from cookielib import CookieJar
from HTMLParser import HTMLParser
from email.mime.text import MIMEText
from email.header import Header

IDENTIFIERTYPE = '1'
IDENTIFIER = 'nnnnnnnnn'
SURNAME = 'XXXXXX'
DOB = 'YYYY-MM-DD'

Sender = 'cicstatus@163.com'
Receiver = 'your@company.com'
SMTPServer = 'smtp.163.com'
SMTPUID = 'cicstatus'
SMTPPWD = 'cic_status'
Subject = 'Application Status as of ' + datetime.datetime.now().strftime("%Y-%m-%d %H:%M")

Result = ''

BaseURL = 'https://services3.cic.gc.ca/ecas/'
URL0 = 'authenticate.do'

class hp(HTMLParser):
    a_text = False

    def handle_starttag(self, tag, attr):
        if tag == 'li':
            if len(attr) == 0: pass
            else:
                for (variable, value) in attr:
                    if variable == "class" and value == "margin-bottom-medium":
                        self.a_text = True

    def handle_endtag(self, tag):
        if tag == 'li':
            self.a_text = False

    def handle_data(self, data):
        global Result
        if self.a_text and len(data.strip()) > 0:
            Result = Result + data.strip() + '\n'

# Get Cookie
Cookie = cookielib.CookieJar()
Opener = urllib2.build_opener(urllib2.HTTPCookieProcessor(Cookie))
Response = Opener.open(BaseURL + URL0)
time.sleep(random.randint(3, 8))

# Login to view Application Status, and get Application Details URL
Request = urllib2.Request(BaseURL + URL0)
FormData = { "lang": "", "_page": "_target0", "app": "", "identifierType": IDENTIFIERTYPE, "identifier": IDENTIFIER, "surname": SURNAME, "dateOfBirth": DOB, "countryOfBirth": "202", "_submit": "Continue" }
DataEncoded = urllib.urlencode(FormData)
Response = Opener.open(Request, DataEncoded)
Content = Response.read()
URL1 = re.search('viewcasehistory.do.*lang=en', Content).group(0).replace("&amp;", "&")
time.sleep(random.randint(3, 8))

# View Application Details by access URL1
Response = Opener.open(BaseURL + URL1)
Content = Response.read()

# HTML Parser
Output = hp()
Output.feed(Content)
Output.close()

print "Current query result length is " + str(len(Result)) + ", modify following condation if necessary."

if len(Result) != 131:
    MSG = MIMEText(Result, 'text', 'utf-8')
    MSG['Subject'] = Header(Subject, 'utf-8')
    SMTP = smtplib.SMTP() #SMTP.set_debuglevel(1)
    SMTP.connect(SMTPServer)
    SMTP.login(SMTPUID, SMTPPWD)
    SMTP.sendmail(Sender, Receiver, MSG.as_string())
    SMTP.quit()

print Result

File: CICStatus.rb
require 'net/http'
require 'net/smtp'
require 'uri'
require 'nokogiri'

_IDENTIFIERTYPE = '1'
_IDENTIFIER = 'NNNNNNNN'
_SURNAME = 'XXXXXX'
_DOB = 'YYYY-MM-DD'

sender = 'yourname@company.com'
receiver = 'yourname@company.com'
smtpServer = 'SMTP.company.com'

_BaseURL = 'https://services3.cic.gc.ca/ecas/'
_URL0 = 'authenticate.do'

uri = URI(_BaseURL + _URL0)
http = Net::HTTP.new(uri.host, uri.port)
http.use_ssl = true

# GET request -> so the host can set his cookies
resp = http.get(uri.path, nil)
cookie = resp.response['set-cookie'].split('; ')[0]

# POST request -> logging in
data = 'lang=&_page=_target0&app=&identifierType=' + _IDENTIFIERTYPE + '&identifier=' + _IDENTIFIER + '&surname=' + _SURNAME + '&dateOfBirth=' + _DOB + '&countryOfBirth=202&_submit=Continue'
headers = {
    'Cookie' => cookie,
    'Referer' => _BaseURL + _URL0,
    'Content-Type' => 'application/x-www-form-urlencoded'
}
resp = http.post(uri.path, data, headers)

# Extract Redirect URL
$_URL1 = ''
resp.each { |key, value|
  if key == 'location' then
    $_URL1 = value
  end }

# POST request -> so we can get details link
uri = URI($_URL1)
resp = http.post(uri.path, data, headers)
$_URL2 = ''
resp.body.each_line { |line|
  if line.include? 'viewcasehistory.do'
    $_URL2 = (line.split("=\""))[1].split("\">")[0].gsub(/&amp;/, '&')
  end
}

# POST request -> for final result
uri = URI(_BaseURL + $_URL2)
data = $_URL2.split('?')[1]
resp = http.post(uri.path, data, headers)

=begin
# DEBUG ONLY
TEMPstr = IO.read("./OUTPUT")
puts TEMPstr.length;
=end

html_doc = Nokogiri::HTML(resp.body)
$_Result = ''
html_doc.css('li').each do |link|
  if "#{link['class']}" == 'margin-bottom-medium'
    $_Result += "#{link.content}\n"
  end
end

puts 'Current query result length is ' + $_Result.length.to_s + ", modify following condation if necessary.\n"
puts $_Result

#message = ''
message = "From: Your Name <" + sender + ">\n"
message += "To: Your Name <" + receiver + ">\n"
message += "Subject: Application Status as of " + Time.new.strftime("%Y-%m-%d %H:%M")+ "\n\n"
message += $_Result

Net::SMTP.start(smtpServer) do |smtp|
  smtp.send_message message, sender, receiver
end
